File: analyzers/tests/SonarAnalyzer.Rules.Tests/Cases/EncryptionAlgorithmsShouldBeSecure_NetStandard21.cs
using System;
using System.Security.Cryptography;

namespace Tests.Diagnostics
{
    class SecureAesCheck
    {
        AesManaged field1 = new AesManaged(); // Noncompliant
        AesManaged Property1 { get; set; } = new AesManaged(); // Noncompliant

        void CtorSetsAllowedValue(byte[] key1, ReadOnlySpan<byte> key2)
        {
            new AesGcm(key1);
            new AesGcm(key2);
        }

        void CtorSetsNotAllowedValue()
        {
            new AesManaged(); // Noncompliant {{Use secure mode and padding scheme.}}
        }

        void InitializerSetsAllowedValue()
        {
            // none
        }

        void InitializerSetsNotAllowedValue()
        {
            new AesManaged() { Mode = CipherMode.CBC }; // Noncompliant
//          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
            new AesManaged() { Mode = CipherMode.CFB }; // Noncompliant
            new AesManaged() { Mode = CipherMode.CTS }; // Noncompliant
            new AesManaged() { Mode = CipherMode.ECB }; // Noncompliant
            new AesManaged() { Mode = CipherMode.OFB }; // Noncompliant
        }

        void PropertySetsNotAllowedValue()
        {
            var c = new AesManaged(); // Noncompliant
            c.Mode = CipherMode.CBC; // Noncompliant, we will be raising twice
            c.Mode = CipherMode.CFB; // Noncompliant, we will be raising twice
            c.Mode = CipherMode.CTS; // Noncompliant, we will be raising twice
            c.Mode = CipherMode.ECB; // Noncompliant, we will be raising twice
            c.Mode = CipherMode.OFB; // Noncompliant, we will be raising twice
        }

        void PropertySetsAllowedValue(bool foo)
        {
            // none
        }
    }

    class RSAEncryptionPaddingTest
    {
        RSACryptoServiceProvider rsaProvider = new RSACryptoServiceProvider();

        void InvocationSetsAllowedValue(byte[] data, RSAEncryptionPadding padding, RSA genericRSA)
        {
            rsaProvider.Encrypt(data, true); // OAEP Padding is used (second parameter set to true)
            rsaProvider.Decrypt(data, false); // Only raise on Encrypt method
            rsaProvider.Encrypt(fOAEP: true, rgb: data);
            rsaProvider.Encrypt(data, System.Security.Cryptography.RSAEncryptionPadding.OaepSHA1);
            rsaProvider.Encrypt(data, RSAEncryptionPadding.OaepSHA256);
            rsaProvider.TryEncrypt(data, null, RSAEncryptionPadding.OaepSHA256, out var bytesWritten1);
            rsaProvider.Encrypt(padding: padding, data: data); // we don't know which padding is actually used here so we do not raise the issue

            genericRSA.Encrypt(data, RSAEncryptionPadding.OaepSHA256);
            genericRSA.TryEncrypt(data, null, RSAEncryptionPadding.OaepSHA256, out var bytesWritten2);
        }

        void InvocationSetsNotAllowedValue(byte[] data, RSA genericRSA)
        {
            rsaProvider.Encrypt(data, false); // Noncompliant {{Use secure mode and padding scheme.}}
            rsaProvider.Encrypt(fOAEP: false, rgb: data); // Noncompliant
            rsaProvider.Encrypt(data, System.Security.Cryptography.RSAEncryptionPadding.Pkcs1); // Noncompliant
            rsaProvider.Encrypt(data, RSAEncryptionPadding.Pkcs1); // Noncompliant
            rsaProvider.TryEncrypt(data, null, RSAEncryptionPadding.Pkcs1, out var bytesWritten1); // Noncompliant

            genericRSA.Encrypt(data, RSAEncryptionPadding.Pkcs1); // Noncompliant
            genericRSA.TryEncrypt(data, null, RSAEncryptionPadding.Pkcs1, out var bytesWritten2); // Noncompliant
        }
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/EncryptionAlgorithmsShouldBeSecure_NetStandard21.vb
Imports System
Imports System.Security.Cryptography

Namespace Tests.Diagnostics
    Class SecureAesCheck
        Private field1 As AesManaged = New AesManaged() ' Noncompliant
        Private Property Property1 As AesManaged = New AesManaged() ' Noncompliant

        Private Sub CtorSetsAllowedValue(ByVal key As Byte())
            Dim aesGcm1 = New AesGcm(key)
        End Sub

        Private Sub CtorSetsNotAllowedValue()
            Dim aesManaged = New AesManaged() ' Noncompliant {{Use secure mode and padding scheme.}}
        End Sub

        Private Sub InitializerSetsAllowedValue()
            ' none
        End Sub

        Private Sub InitializerSetsNotAllowedValue()
            Dim aesManaged = New AesManaged() With {.Mode = CipherMode.CBC} ' Noncompliant
'                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
            aesManaged = New AesManaged() With {.Mode = CipherMode.CFB} ' Noncompliant
            aesManaged = New AesManaged() With {.Mode = CipherMode.CTS} ' Noncompliant
            aesManaged = New AesManaged() With {.Mode = CipherMode.ECB} ' Noncompliant
            aesManaged = New AesManaged() With {.Mode = CipherMode.OFB} ' Noncompliant
        End Sub

        Private Sub PropertySetsNotAllowedValue()
            Dim c = New AesManaged() ' Noncompliant
            c.Mode = CipherMode.CBC ' Noncompliant
            c.Mode = CipherMode.CFB ' Noncompliant
            c.Mode = CipherMode.CTS ' Noncompliant
            c.Mode = CipherMode.ECB ' Noncompliant
            c.Mode = CipherMode.OFB ' Noncompliant
        End Sub

        Private Sub PropertySetsAllowedValue(ByVal foo As Boolean)
            ' none
        End Sub
    End Class

    Class RSAEncryptionPaddingTest
        Private rsaProvider As RSACryptoServiceProvider = New RSACryptoServiceProvider()

        Private Sub InvocationSetsAllowedValue(ByVal data As Byte(), ByVal padding As RSAEncryptionPadding, ByVal genericRSA As RSA)
            rsaProvider.Encrypt(data, True)
            rsaProvider.Decrypt(data, False)
            rsaProvider.Encrypt(fOAEP:=True, rgb:=data)
            rsaProvider.Encrypt(data, System.Security.Cryptography.RSAEncryptionPadding.OaepSHA1)
            rsaProvider.Encrypt(data, RSAEncryptionPadding.OaepSHA256)
            rsaProvider.Encrypt(padding:=padding, data:=data)
        End Sub

        Private Sub InvocationSetsNotAllowedValue(ByVal data As Byte(), ByVal genericRSA As RSA)
            rsaProvider.Encrypt(data, False) ' Noncompliant
            rsaProvider.Encrypt(fOAEP:=False, rgb:=data) ' Noncompliant
            rsaProvider.Encrypt(data, System.Security.Cryptography.RSAEncryptionPadding.Pkcs1) ' Noncompliant
            rsaProvider.Encrypt(data, RSAEncryptionPadding.Pkcs1) ' Noncompliant
        End Sub
    End Class
End Namespace
